bitnami/redis/templates/scripts-configmap.yaml

```
@@ -64,10 +64,15 @@ data:
 
     SENTINEL_SERVICE_PORT=$(get_port "{{ include "common.names.fullname" . }}" "TCP_SENTINEL")
     validate_quorum() {
+        redis_service=${REDIS_SERVICE}
+        if ! timeout 1 bash -c "cat < /dev/null > /dev/tcp/${REDIS_SERVICE}/${SENTINEL_SERVICE_PORT}"; then
+            warn "Timeout when connecting to ${REDIS_SERVICE}, will use ${HEADLESS_SERVICE}"
+            redis_service=${HEADLESS_SERVICE}
+        fi
```

`$REDIS_SERVICE` could be an external master (`$REDIS_EXTERNAL_MASTER_HOST`) in that case doesn't make sense to use the headless service.

I think we could use directly the headless service in this function unless we set the external master host. Here is my proposal:

```suggestion
        redis_service=${REDIS_EXTERNAL_MASTER_HOST:-$HEADLESS_SERVICE}
```
I think there is no need to use the `timeout` validation because the headless service will publish the non ready endpoints due to its configuration.

Could you give it a try? I'll open an internal question to ensure this is correct

---

For the master (replicas 0) it is always unavailable (because it is starting up). Therefore, it is advisable to start via headless to get a clean reused connection instead of pointless waiting.
For slaves (or replicas greater than 0), it should only start when it finds an available connection through the service, not the headless service. This will help the redis cluster to only accept new members when the cluster is up and running

---

Sorry @nobiit for my very late response

I think we had a little misunderstood here (or my head is a little bit messy, that's another option). I am going to try to do a recap:

#### Current Situation

I will be very schematic and I will assume `redis` as release name:
* Pods are managed by a `statefulset`.
* redis service will use all pods from created by the `statefulset`.
* redis-headless service is a headless and it won't wait till readiness tests to expose the endpoints. It uses the same pods as the redis service.
* Each pod has 2 containers: redis and sentinel. 
* Redis container in first pod (`redis-node-0`) is configured to be master.  AFAIK It doesn't have connections to redis service.
* Sentinel container in first pod (`redis-node-0`) tries to connect to redis service to get master info. According to your logs, in this point the release is getting stuck, the statefulset doesn't create the following pod.

#### Your Solution

Change `validate_quorum` function used by redis 'replica' containers  (redis-node-1, redis-node-2, etc) during startup.


I don't fully understand how your solution fix the problem (probably I am missing something). I'll try to setup a cluster with Cillium, but I'll need some time

---

Sorry I have no chance to test it.

@nobiit I don't fully understand how changing the `validate_quorum` function used by the replica containers (redis-node-1, redis-node-2, etc) fixes an issue that is appearing in redis-node-0. I think I am missing something